src/Native/Unix/System.Native/pal_process.c

```diff
@@ -233,15 +237,74 @@ int32_t SystemNative_ForkAndExecProcess(const char* filename,
+    // need to set all writable pages in the parent process to copy-on-write because the child process
+    // is allowed to write to the parent process memory pages.
+
+    // The thing to remember about shared memory vfork() is the documentation is way out of date.
+    // It does the following things:
+    // * creates a new process in the memory space of the calling process.
+    // * blocks the calling thread (not process!) in an uninterruptable sleep
+    // * sets up the process records so the following happen:
+    //   + execve() replaces the memory space in the child and unblocks the parent
+    //   + process exit by any means unblocks the parent
+    //   + ptrace() makes a security demand against the parent process
+    //   + accessing the terminal with read() or write() fail in system-dependent ways
+    // Due to lack of documentation, setting signal handlers in the vfork() child is a bad idea. We don't
+    // do this, but it's worth pointing out.
+
+    // All platforms that provide shared memory vfork() check the parent process's context when
+    // ptrace() is used on the child, thus making setuid() safe to use after vfork(). The fabled vfork()
+    // security hole is the other way around; if a multithreaded host were to execute setuid()
+    // on another thread while a vfork() child is still pending, bad things are possible; however we
+    // do not do that.
+
+    if ((processId = vfork()) == 0) // processId == 0 if this is child process
+#else
+    if ((processId = fork()) == 0) // processId == 0 if this is child process
+#endif
     {
-        success = false;
-        goto done;
-    }
+        // It turns out that child processes depend on their sigmask being set to something sane rather than mask all.
+        // On the other hand, we have to mask all to avoid our own signal handlers running in the child process, writing
+        // to the pipe, and waking up the handling thread in the parent process. This also avoids third-party code getting
+        // equally confused.
+        // Remove all signals, then restore signal mask.
+        // Since we are in a vfork() child, the only safe signal values are SIG_DFL and SIG_IGN.  See man 3 libthr on BSD.
+        // "The implementation interposes the user-installed signal(3) handlers....to pospone signal delivery to threads
+        // which entered (libthr-internal) critical sections..."  We want to pass SIG_DFL anyway.
+        sigset_t junk_signal_set;
+        struct sigaction sa_default;
+        struct sigaction sa_old;
+        memset(&sa_default, 0, sizeof(sa_default)); // On some architectures, sa_mask is a struct so assigning zero to it doesn't compile
+        sa_default.sa_handler = SIG_DFL;
+        for (int sig = 1; ; ++sig)
+        {
+            if (sig == SIGKILL || sig == SIGSTOP)
+            {
+                continue;
+            }
+            if (sigaction(sig, NULL, &sa_old))
+            {
+                break; // No more signals
```

Comment:
Breaking here is incorrect because `sigaction` in glibc [returns `-1`](https://sourceware.org/git/?p=glibc.git;a=blob;f=nptl/sigaction.c;h=3c0438ed44616c4ef5e15db1dc2d412acf0bc8da;hb=HEAD#l24) for signals used by it internally. Since in practice those signals are 32/33 and the total number of signals is >=64, the loop will stop early.

Reply:
It's documented: http://man7.org/linux/man-pages/man2/sigaction.2.html (section C library/kernel differences).

Reply:
I really don't want to cause any controversy on this; but somehow my man page on sigaction has about half the content. I guess they fixed their docs.